libraries/AC_WPNav/AC_WPNav.cpp

```diff
@@ -647,16 +559,49 @@ void AC_WPNav::set_yaw_cd(float heading_cd)
     _flags.wp_yaw_set = true;
 }
 
+// set yaw rate used for spline and waypoint navigation
+void AC_WPNav::set_yaw_rate_cds(float yaw_rate_cds)
+{
+    _yaw_rate = yaw_rate_cds;
+}
+
+// get terrain's altitude (in cm above the ekf origin) at the current position (+ve means terrain below vehicle is above ekf origin's altitude)
+bool AC_WPNav::get_terrain_offset(float& offset_cm)
+{
+    // calculate offset based on source (rangefinder or terrain database)
+    switch (get_terrain_source()) {
+    case AC_WPNav::TerrainSource::TERRAIN_UNAVAILABLE:
+        return false;
+    case AC_WPNav::TerrainSource::TERRAIN_FROM_RANGEFINDER:
+        if (_rangefinder_healthy) {
+            offset_cm = _inav.get_altitude() - _rangefinder_alt_cm;
+            return true;
+        }
+        return false;
+    case AC_WPNav::TerrainSource::TERRAIN_FROM_TERRAINDATABASE:
+#if AP_TERRAIN_AVAILABLE
+        float terr_alt = 0.0f;
+        if (_terrain != nullptr && _terrain->height_above_terrain(terr_alt, true)) {
+            offset_cm = _inav.get_altitude() - (terr_alt * 100.0f);
+            return true;
+        }
+#endif
+        return false;
+    }
+
+    // we should never get here but just in case
+    return false;
+}
+
 ///
 /// spline methods
 ///
 
 /// set_spline_destination waypoint using location class
 ///     returns false if conversion from location to vector from ekf origin cannot be calculated
-///     stopped_at_start should be set to true if vehicle is stopped at the origin
-///     seg_end_type should be set to stopped, straight or spline depending upon the next segment's type
-///     next_destination should be set to the next segment's destination if the seg_end_type is SEGMENT_END_STRAIGHT or SEGMENT_END_SPLINE
-bool AC_WPNav::set_spline_destination(const Location& destination, bool stopped_at_start, spline_segment_end_type seg_end_type, Location next_destination)
+///     next_destination should be set to the next segment's destination
+///     next_is_spline should be true if path to next_destination should be a spline
+bool AC_WPNav::set_spline_destination_loc(const Location& destination, Location next_destination, bool next_is_spline)
```

Comment:
txs.  next_destination is now passed as const.  We couldn't do this in master because within the function (in master) we change the altitude frame but we can't do that with s-scurves.